### PR TITLE
Modify ClientPrefs to be much more user-accessible

### PR DESCRIPTION
**Pull request by Starveldt**
I did a bunch of rewriting to ClientPrefs, added a new function called "getPref" to get a certain ClientPref (with a default value like how getGameplaySetting does it), and hopefully made it a little more lua-friendly.

It should also be completely compatible with old ClientPrefs; I only took all of the settings and put them in a map, plus separated load functions.

This is also my first PR, so be nice on me please :)

### CODE REVIEW

**Review by iym1337 — APPROVED** [review]

**Review by EyeDaleHim — APPROVED** [review]
great job mate, fnf week 7 also uses this same approach

**Comment by BeastlyGhost** [comment]
feels like an interesting thing to have, clientprefs rn is pretty messy in my opinion, glad someone did this already

looking forward for this one!

**Comment by Paracosm-Daemon** [comment]
> feels like an interesting thing to have, clientprefs rn is pretty messy in my opinion, glad someone did this already
> 
> looking forward for this one!

thanks haha, i actually was expecting to not be the first, but as it turns out i am, which is surprising--i actually have used a method that's similar for one of my previous projects

**Review by Raltyro — COMMENTED** [review]
Would approve but what would happen to the luas that uses getPropertyFromClass("ClientPrefs") ?

**Comment by Paracosm-Daemon** [comment]
> Would approve but what would happen to the luas that uses getPropertyFromClass("ClientPrefs") ?

that's a good question, i suppose it'd be obselete since it's now softcoded instead of you being able to just get the pref directly from the class, though in turn you can make a new lua function called 'getPref' and just replace all mentions of 'getPropertyFromClass' with 'getPref', though it doesn't really support 'legacy' scripts.

a way to mitigate this, i believe, would be to check if the lua script is trying to get a setting from ClientPrefs inside of the 'getPropertyFromClass' function, then try to get a pref, i.e

```hx
Lua_helper.add_callback(lua, "getPropertyFromClass", function(classVar:String, variable:String) {
	@:privateAccess
	var killMe:Array<String> = variable.split('.');
	if(killMe.length > 1) {
		var coverMeInPiss:Dynamic = getVarInArray(Type.resolveClass(classVar), killMe[0]);
		for (i in 1...killMe.length-1) {
			coverMeInPiss = getVarInArray(coverMeInPiss, killMe[i]);
		}
		return getVarInArray(coverMeInPiss, killMe[killMe.length-1]);
	}
	switch (classVar) // Legacy support
	{
		case 'ClientPrefs':
		{
			var pref:Dynamic = ClientPrefs.getPref(variable);
			if (pref != null) return pref;
		}
	}
	return getVarInArray(Type.resolveClass(classVar), variable);
});
```
this should hopefully fix it, and i think i'll actually implement this right now, sorry for not thinking about this lol

File: source/FunkinLua.hx [hunk 1]
```diff
@@ -697,6 +697,10 @@ class FunkinLua {
 			}
 		});
 
+		Lua_helper.add_callback(lua, "getPref", function(pref:String, ?defaultValue:Dynamic) {
+			return ClientPrefs.getPref(pref, defaultValue);
+		});
+
```
Comment (ImaginationSuperHero528) [inline source/FunkinLua.hx:703]: Is there gonna be setPref?

**Review by rodney528 — APPROVED** [review]

File: source/FunkinLua.hx (hunk 1 above)
Comment (Paracosm-Daemon) [reply source/FunkinLua.hx:703]: > Is there gonna be setPref?

yeah, my bad--i was in a rush, but i can make that soon

**Review by Starveldt — COMMENTED** [review]
can i use field/setField without screwing anything up? i'm using get/setproperty to be safe, but it may have a performance impact, which i'm worried about

**Review by rodney528 — APPROVED** [review]

**Review by rodney528 — APPROVED** [review]

**Comment by BeastlyGhost** [comment]
notice: images persist is not an option anymore, as it was "removed" in favor of a better and more efficient paths system

**Comment by Paracosm-Daemon** [comment]
> notice: images persist is not an option anymore, as it was "removed" in favor of a better and more efficient paths system

i know, thats why i commented it out

**Review by rodney528 — APPROVED** [review]

**Review by DM-kun — APPROVED** [review]

**Review by rodney528 — APPROVED** [review]

**Review by MemeHoovy — APPROVED** [review]
Alright.

**Review by rodney528 — APPROVED** [review]

**Comment by Paracosm-Daemon** [comment]
> Alright, but it seems like a huge pain in the ass for those who port it over to their mod manually though, just saying.

just wanted to say, i can understand this, but it's actually not that hard from source--all i did in vscode was ctrl + shift + h and just go through every setting and replace with ClientPrefs.getPref('prefname') and that was about it, so it shouldn't be too bad hopefully

**Review by rodney528 — APPROVED** [review]

**Comment by MemeHoovy** [comment]
> 

I see.

**Comment by MemeHoovy** [comment]
> > Alright, but it seems like a huge pain in the ass for those who port it over to their mod manually though, just saying.
> 
> just wanted to say, i can understand this, but it's actually not that hard from source--i actually did this for my mod manually. in vscode, all i did was ctrl + shift + h and just go through every setting and replace ClientPrefs.prefname with ClientPrefs.getPref('prefname') and that was about it, so it shouldn't be too bad hopefully

I see.

**Comment by BeastlyGhost** [comment]
I should prob merge this to some private build, seems easier to use

**Review by rodney528 — APPROVED** [review]

**Review by Intancote — COMMENTED** [review]
Would approve if there were no merge conflicts

**Comment by Paracosm-Daemon** [comment]
> Would approve if there were no merge conflicts

ty for telling me, i'll fix those rn

**Review by Intancote — APPROVED** [review]
👍

**Review by JorosTheModderPH — APPROVED** [review]

**Review by rodney528 — APPROVED** [review]

**Review by Parsellol — APPROVED** [review]

**Review by rodney528 — APPROVED** [review]

**Review by rodney528 — APPROVED** [review]

**Review by Parsellol — APPROVED** [review]

**Review by rodney528 — APPROVED** [review]

**Review by rodney528 — APPROVED** [review]

**Review by rodney528 — APPROVED** [review]

**Review by rodney528 — APPROVED** [review]

**Comment by ImaginationSuperHero528** [comment]
> > Would approve but what would happen to the luas that uses getPropertyFromClass("ClientPrefs") ?
> 
> that's a good question, i suppose it'd be obselete since it's now softcoded instead of you being able to just get the pref directly from the class, though in turn you can make a new lua function called 'getPref' and just replace all mentions of 'getPropertyFromClass' with 'getPref', though it doesn't really support 'legacy' scripts.
> 
> a way to mitigate this, i believe, would be to check if the lua script is trying to get a setting from ClientPrefs inside of the 'getPropertyFromClass' function, then try to get a pref, i.e
> 
> ```hx
> Lua_helper.add_callback(lua, "getPropertyFromClass", function(classVar:String, variable:String) {
> 	@:privateAccess
> 	var killMe:Array<String> = variable.split('.');
> 	if(killMe.length > 1) {
> 		var coverMeInPiss:Dynamic = getVarInArray(Type.resolveClass(classVar), killMe[0]);
> 		for (i in 1...killMe.length-1) {
> 			coverMeInPiss = getVarInArray(coverMeInPiss, killMe[i]);
> 		}
> 		return getVarInArray(coverMeInPiss, killMe[killMe.length-1]);
> 	}
> 	switch (classVar) // Legacy support
> 	{
> 		case 'ClientPrefs':
> 		{
> 			var pref:Dynamic = ClientPrefs.getPref(variable);
> 			if (pref != null) return pref;
> 		}
> 	}
> 	return getVarInArray(Type.resolveClass(classVar), variable);
> });
> ```
> 
> this should hopefully fix it, and i think i'll actually implement this right now, sorry for not thinking about this lol

Does setPropertyFromClass have this check?

**Comment by Paracosm-Daemon** [comment]
> > > Would approve but what would happen to the luas that uses getPropertyFromClass("ClientPrefs") ?
> 
> > 
> 
> > that's a good question, i suppose it'd be obselete since it's now softcoded instead of you being able to just get the pref directly from the class, though in turn you can make a new lua function called 'getPref' and just replace all mentions of 'getPropertyFromClass' with 'getPref', though it doesn't really support 'legacy' scripts.
> 
> > 
> 
> > a way to mitigate this, i believe, would be to check if the lua script is trying to get a setting from ClientPrefs inside of the 'getPropertyFromClass' function, then try to get a pref, i.e
> 
> > 
> 
> > ```hx
> 
> > Lua_helper.add_callback(lua, "getPropertyFromClass", function(classVar:String, variable:String) {
> 
> > 	@:privateAccess
> 
> > 	var killMe:Array<String> = variable.split('.');
> 
> > 	if(killMe.length > 1) {
> 
> > 		var coverMeInPiss:Dynamic = getVarInArray(Type.resolveClass(classVar), killMe[0]);
> 
> > 		for (i in 1...killMe.length-1) {
> 
> > 			coverMeInPiss = getVarInArray(coverMeInPiss, killMe[i]);
> 
> > 		}
> 
> > 		return getVarInArray(coverMeInPiss, killMe[killMe.length-1]);
> 
> > 	}
> 
> > 	switch (classVar) // Legacy support
> 
> > 	{
> 
> > 		case 'ClientPrefs':
> 
> > 		{
> 
> > 			var pref:Dynamic = ClientPrefs.getPref(variable);
> 
> > 			if (pref != null) return pref;
> 
> > 		}
> 
> > 	}
> 
> > 	return getVarInArray(Type.resolveClass(classVar), variable);
> 
> > });
> 
> > ```
> 
> > 
> 
> > this should hopefully fix it, and i think i'll actually implement this right now, sorry for not thinking about this lol
> 
> 
> 
> Does setPropertyFromClass have this check?

yeah it should

**Review by rodney528 — APPROVED** [review]
:D

**Review by Raltyro — APPROVED** [review]
came to check and it looks better than ever

**Review by libauror-a — APPROVED** [review]
aww man if this gets accepted ill have to remove the coment of psych emgine clientprefs sucking in my engine, good pr tho

**Comment by Paracosm-Daemon** [comment]
> aww man if this gets accepted ill have to remove the coment of psych emgine clientprefs sucking in my engine, good pr tho

idk if this'll get accepted since i like. massively rewrote clientprefs as a whole, @ShadowMario can decide on that, but this is good if you wanna modify psych engine yourself. so you can still keep that comment if you don't plan on adding this pr lol

**Review by YanniZ06 — APPROVED** [review]
Finally

**Pull request closed without merging**